File: Sorting/QuickSort/quick_sort_hoare.py
# Implementation of Quick sort using Hoare's partition

def partition_hoare(lo, hi, array: list) -> int:
    pivot_index = hi
    pivot = array[pivot_index]
    hi -= 1

    while(True):
        while array[lo] < pivot:
            lo += 1
        while array[hi] > pivot:
            hi -= 1

        if lo >= hi:
            break
        else:
            array[lo], array[hi] = array[hi], array[lo]
            lo += 1
    array[pivot_index], array[lo] = array[lo], array[pivot_index]
    print(f"Final array = {array} and lo value = {lo}")
    return lo 

def quick_sort(lo, hi, array: list) -> list:
    if hi - lo <= 0:
        return
    partition_index = partition_hoare(lo, hi, array)
    #quick_sort(lo, partition_index - 1, array)
    #quick_sort(partition_index + 1, hi, array)
    return array

array = [5, 3, 1, 7, 6, 9, 10, 2, 4, 8]
print("Unsorted array: ")
print(array)

print("Sorted array: ") 
print(quick_sort(0, len(array) - 1, array))

File: Sorting/QuickSort/QuickSortHoare.js
/*
 * Quick sort implementation using Hoare partition scheme.
 *
 * */

function partitionHoare(lo, hi, array){
	let pivotIndex = hi;
	let pivot = array[pivotIndex];
	hi--;

	while(true){
		while(array[lo] < pivot) lo++;
		while(array[hi] > pivot) hi--;
		if(lo >= hi){
			break;
		} else{
			let temp = array[lo];
			array[lo] = array[hi];
			array[hi] = temp;
			lo++;
		}
	}

	let temp = array[lo];
	array[lo] = array[pivotIndex];
	array[pivotIndex] = temp;
	
	return lo;
}

function quickSort(lo, hi, array){
	if(hi - lo <= 0) return;
	let partitionIndex = partitionHoare(lo, hi, array);
	quickSort(lo, partitionIndex - 1, array);
	quickSort(partitionIndex + 1, hi, array);
	return array;
}

let array = [5, 3, 1, 7, 6, 9, 10, 2, 4, 8];

console.log("Unsorted array: ");
console.log(array)

console.log("Sorted array: ");
console.log(quickSort(0, array.length - 1, array))
